android/src/com/mapswithme/maps/traffic/TrafficManager.java

```diff
@@ -0,0 +1,181 @@
+package com.mapswithme.maps.traffic;
+
+import android.support.annotation.MainThread;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.mapswithme.util.Utils;
+import com.mapswithme.util.log.DebugLogger;
+import com.mapswithme.util.log.Logger;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public enum TrafficManager
+{
+  INSTANCE;
+  @NonNull
+  private final Logger mLogger = new DebugLogger(TrafficManager.class.getSimpleName());
+  @NonNull
+  private final TrafficState.StateChangeListener mStateChangeListener = new TrafficStateListener();
+  @TrafficState.Value
+  private int mState = TrafficState.DISABLED;
+  @NonNull
+  private final List<TrafficCallback> mCallbacks = new ArrayList<>();
+
+  TrafficManager()
+  {
+    mLogger.d("Traffic manager initialization");
+    setCoreStateChangedListener();
+  }
+
+  private void setCoreStateChangedListener()
+  {
+    mLogger.d("Set core traffic state listener");
+    TrafficState.nativeSetListener(mStateChangeListener);
+  }
+
+  public void enableOrDisable()
+  {
+    if (mState == TrafficState.DISABLED)
+    {
+      enable();
+      return;
+    }
+
+    disable();
+  }
+
+  private void enable()
+  {
+    mLogger.d("Enable traffic");
+    TrafficState.nativeEnable();
+  }
+
+  public void disable()
+  {
+    mLogger.d("Disable traffic");
+    TrafficState.nativeDisable();
+  }
+
+  public boolean isEnabled()
+  {
+    return mState != TrafficState.DISABLED;
+  }
+
+  public void attach(@NonNull TrafficCallback callback)
+  {
+    if (mCallbacks.contains(callback))
+    {
+      throw new IllegalStateException("A callback '" + callback
+                                      + "' is already attached. Check that the 'detach' method was called.");
+    }
+
+    mLogger.d("Attach callback '" + callback + "'");
+    mCallbacks.add(callback);
+    postPendingState();
+  }
+
+  private void postPendingState()
+  {
+    mStateChangeListener.onTrafficStateChanged(mState);
+  }
+
+  public void detach()
+  {
+    if (mCallbacks.isEmpty())
+    {
+      throw new IllegalStateException("There are no attached callbacks. Invoke the 'detach' method " +
+                                      "only when it's really needed!");
+    }
+
+    for (TrafficCallback callback : mCallbacks)
+      mLogger.d("Detach callback '" + callback + "'");
+    mCallbacks.clear();
+  }
+
+  private class TrafficStateListener implements TrafficState.StateChangeListener
+  {
+    @Override
+    @MainThread
+    public void onTrafficStateChanged(@TrafficState.Value int state)
+    {
+      mLogger.d("onTrafficStateChanged current state = " + TrafficState.nameOf(mState)
+                + " new value = " + TrafficState.nameOf(state));
+      mState = state;
+
+      if (mCallbacks.isEmpty())
```

Comment:
No need in this check. iterateOverCallbacks() is a no-op when there are no callbacks.